[Eslint Tools] add shared @kbn/eslint-rule-overrides package

kapral18 opened the pull request:
## 🎯 Summary

Introduces `@kbn/eslint-rule-overrides` - a utility package that enables teams to safely inherit and customize ESLint rules from Kibana's root configuration in nested directory contexts.

Addresses #223242

## 🔧 Problem Statement

Kibana's root `.eslintrc.js` is **2400+ lines** with rules spread across multiple overrides. 

Making targeted changes is:
- **Risky**: Easy to accidentally affect other teams' configurations
- **Complex**: Requires calculating interactions between multiple override blocks  
- **Fragile**: Small changes can have unexpected cascading effects
- **Time-consuming**: Understanding the full impact requires analyzing the entire file

### Real-World Example
Our team needed to ban Enzyme imports during our RTL migration, so that we could iteratively migrate tests to RTL with multiple PRs while restricting others (including contributors from outside the team) from contributing new enzyme tests to our folders. This was a team decision and it happened before global initiative to deprecate enzyme on kibana level.

We tried to add the bans to global config but it turned out to be more of a debugging session than a config change. 

And in the meanwhile:
- Full migration took multiple weeks and is still not fully done to this moment (only first sub-team folders have been migrated)
- We own 20+ folders scattered across `src/`, `x-pack/`, and `packages/` so detecting and persisting path in global eslint config in itself was a challenge
- Adding folder-specific rules to root config for each PR would involve going through operations team for every PR
- Reverting changes after migration would involve the same amount of effort
- And there was always a chance of messing up stuff and breaking rules for some other team

That's why initially we chose to just [introduce](https://github.com/elastic/kibana/pull/218134/files#diff-8c483e89533bf9ec9c731e02de23b77255f4d6b048dcbd4acbc9b37e7794cfbc) multiple local team specific .eslintrc.js files taking advantage of eslint's [cascade](https://eslint.org/docs/latest/use/configure/configuration-files-deprecated#cascading-and-hierarchy) mechanic which served us well. Even tho was a little ugly.

### Real-World Example 2

A little later we did an investigation into test performance and [found out](https://github.com/elastic/kibana/pull/223838) that some accessibility selectors of RTL are having substantial performance impact on jest tests especially with deep render trees and we decided to ban them for our team.

And we added new import bans to those same child config files with ease.

### Still not good enough

But clearly these files and this solution felt hacky, because it was all over the place so we needed to centralize it somewhere. 

And since we didn't have a good place to store it I took the opportunity and reached out to @delanni to suggest to move this solution into some kibana level shared place and generalize it to be potentially helpful to others with the similar need.

That's how this PR came about.

## ✨ Solution

This package provides a **safe, tested abstraction** for extending root ESLint rules locally:

```javascript
// Before: Risky root config modification affecting everyone
// After: Safe local override in your team's directory
const { createRuleOverrides } = require('@kbn/eslint-rule-overrides');

module.exports = {
  overrides: createRuleOverrides({
    childConfigDir: __dirname,
    rules: {
      // Simple severity change
      'no-console': { severity: 'error' },
      
      // Merge additional restrictions
      'no-restricted-imports': {
        strategy: 'merge',
        value: ['enzyme', 'react-dom/test-utils']
      }
    }
  })
};
```

## 🚀 Key Features

### 1. **Intelligent Rule Inheritance**
- Automatically inherits applicable rules from root config
- Transforms file patterns to work correctly in nested contexts
- Handles complex glob patterns and negations

### 2. **Safe Operations by Default**
- Rule severity changes work for any rule
- Replace/remove strategies for explicit overrides
- Merge operations require explicit handlers to prevent corruption

### 3. **Extensible Architecture** (Check README)
- Built-in handler for `no-restricted-imports` (with intelligent merging)
- Framework for adding handlers for other complex rules
- Custom handler support for inline experimentation

### 4. **Test Coverage**
- Comprehensive test suite
- Clear error messages for debugging
- Full documentation with examples

## 📊 Benefits

| Aspect | Without Package | With Package |
|--------|----------------|--------------|
| **Risk** | Modify 2500+ line file affecting all teams | Local changes only affect your directory |
| **Complexity** | Calculate override interactions manually | Package handles pattern transformation |
| **Iteration Speed** | Careful PR review for each change in god config file | Quick, safe iterations in local child eslint files |
| **Rollback** | Complex reversion process | Just delete local overrides or the entire `.eslintrc.js` |
| **Testing** | Hard to test in isolation | Local config easy to validate |
| **File Path Migrations** | Root eslint config hardcodes deep paths, that are prone to change every time folders are moved around during migrations | Folder scoped nested .eslintrc.js files use config relative paths so that problem doesn't exist |

## Drawbacks

- Heavily reliant on `v8` cascade mechanics. Potentially an impediment for smooth upgrade to `v9`. This needs to be adapted to new flat config.
- `customHandler` is not simple and requires package interface knowledge atm. Can be simplified in the future tho and supporting documentation is hopefully elaborate and helpful. Accompanying tests are plentiful.
- Currently only a single elaborate builtin handler is available. For `no-restricted-imports` rule.  Other similarly complex rules should be added and maintained by contributors moving forward separately. So the process of adding new complex rules is not exactly fast atm.
- Child `.eslintrc.js` files are adding to fragmentation of linting configs across codebase which can be a problem from operations/platform team perspective with regards to equal rule enforcement
- Webstorm has a bug which involves manually setting working directory of eslint in settings to kibana root (once). [Explained here](https://github.com/elastic/kibana/pull/227301) which is easily resolvable. The problem is related not to package but to usage of nested .eslintrc.js config file in general.

## 🤝 Ownership & Maintenance

- **Owner**: Not clear yet but I am thinking about @elastic/kibana-operations (for broad applicability) or @elastic/security-threat-hunting-investigations (because I was part of that team when created the solution and the solution atm addresses the existing need there)
- **Maintainer**: I can assist with maintenance when I am available.
- **Contributors**: Teams can/should add handlers for their specific rules into this package

## 💡 Why this matters?

- ⏳ Allows to make rules on team level stricter before global ban, which significantly improves team developer ux considering how painfully slow all dev processes are in our gigarepo.⏳
- 🦾 Facilitates easier customization of linting rules on team level which leads to more involvement of teams in improving their own linting and code quality as a whole 🦾
-  ⚠️ Folder-based linting with nested .eslintrc.js files is explicit and allows to easily enforce team A's rules in code that resides in team B's packages/plugins but still belongs to team A, which is a situation that is plenty present in complex meta plugins like security solution. ⚠️
-  🧠 Cognitive load matters. Root config file is only getting more and more complicated, but it's already well above cognitive comprehension limits. It not only slows down and complicates the process of changing linting rules but also scares away and demotivates engineers to engage with linting altogether. 🧠

---

### Try It Out in your team folder's `.eslintrc.js`

```js
echo "const { createRuleOverrides } = require('@kbn/eslint-rule-overrides');

module.exports = {
  overrides: createRuleOverrides({
    childConfigDir: __dirname,
    rules: {
      // Your overrides here
    }
  })
};"
```

```bash
# Test it works
yarn lint:es path/to/your/plugin/or/package/or/file
```

kapral18 committed:
[Eslint tools] add shared @kbn/eslint-rule-overrides package

addresses #223242

- Add new package `@kbn/eslint-rule-overrides` to manage ESLint rule
overrides across the Kibana codebase
- Include README.md with detailed documentation and examples
- Create `jest.config.js` for testing configuration
- Add `kibana.jsonc` with package metadata
- Create `package.json` with package details
- Implement `index.ts` for package exports
- Add `no-restricted-imports.test.ts` for unit tests of the main
functionality
- Implement `no-restricted-imports.ts` with the core logic for creating
overrides
- Define `types.ts` with TypeScript types for the package
- Add `utils.test.ts` for testing utility functions
- Implement `utils.ts` with utility functions for path manipulation and
merging configurations
- Update `tsconfig.base.json` to include new package paths
- Add entry for `@kbn/eslint-rule-overrides` in `yarn.lock`

kapral18 committed:
feat: add kbn-eslint-rule-overrides package and update CODEOWNERS
- Added `kbn-eslint-rule-overrides` package for managing ESLint rule overrides.
- Updated `.github/CODEOWNERS` to include `@elastic/security-threat-hunting-investigations` for the new package.
- Renamed `packages/kbn-eslint-rule-overrides/src/index.ts` to `packages/kbn-eslint-rule-overrides/index.ts`.
- Updated import paths in `packages/kbn-eslint-rule-overrides/index.ts` to reflect the new structure.
- Added `types` field to `packages/kbn-eslint-rule-overrides/package.json` to point to `src/types.ts`.
- Updated `packages/kbn-eslint-rule-overrides/src/no-restricted-imports.test.ts` to include `rootDir` in test cases.
- Refactored `packages/kbn-eslint-rule-overrides/src/no-restricted-imports.ts` to dynamically determine the root directory using `git rev-parse`.
- Simplified and optimized the logic in `packages/kbn-eslint-rule-overrides/src/no-restricted-imports.ts` for creating and merging ESLint overrides.
- Updated `packages/kbn-eslint-rule-overrides/src/types.ts` to import `Linter` from `eslint` and cleaned up types.
- Removed `packages/kbn-eslint-rule-overrides/src/utils.test.ts` and `packages/kbn-eslint-rule-overrides/src/utils.ts` as they were no longer needed.
- Cleaned up `packages/kbn-eslint-rule-overrides/tsconfig.json` for consistency.
- Updated `x-pack/solutions/security/plugins/security_solution/public/onboarding/.eslintrc.js` to use the new `createNoRestrictedImportsOverrides` function from `@kbn/eslint-rule-overrides`.

kapral18 committed:
chore: remove TypeScript files and types

- Rename `index.ts` to `index.js` and update exports
- Rename `no-restricted-imports.ts` to `no-restricted-imports.js` and
update code
- Add `no-restricted-imports.test.js` for testing
- Update README to remove detailed API documentation
- Adjust `.eslintrc.js` usage in example
- Remove `tsconfig.json` and related TypeScript references

kapral18 committed:
fix: update kbn-eslint-rule-overrides ownership and enhance documentation

- Update ownership of `kbn-eslint-rule-overrides` in `CODEOWNERS` and `kibana.jsonc` to `@elastic/kibana-operations`
- Rename `no-restricted-imports.js` to `create_no_restricted_imports_override.js` and update references
- Enhance `README.md` with detailed purpose, installation, usage, and API documentation
- Add comprehensive unit tests in `create_no_restricted_imports_override.test.js` and remove old tests from `no-restricted-imports.test.js`
- Improve error handling and logging in `create_no_restricted_imports_override.js`
- Update `.eslintrc.js` in `x-pack/solutions/security/plugins/security_solution/public/onboarding` to use the new import path

kapral18 committed:
fix: update create_no_restricted_imports_override to use childConfigDir

- Removed dynamic root directory detection using `execSync` and `git rev-parse`.
- Added `childConfigDir` to `CreateOverrideOptions` and require it to be provided.
- Updated `createNoRestrictedImportsOverride` to use `childConfigDir` for context.
- Modified `.eslintrc.js` in `security_solution/public/onboarding` to pass `__dirname` as `childConfigDir`.

kapral18 committed:
fix: resolve root directory dynamically and update tests

- Remove hardcoded `ROOT_DIR` and use `rootDir` variable for resolving paths dynamically.
- Update the mock setup in tests to intercept `path.resolve` calls and return a fixed path for `ROOT_DIR`.
- Replace the custom mock implementation of `create_no_restricted_imports_override` with a simpler approach using `jest.requireActual`.
- Add a test case to ensure an error is thrown when `childConfigDir` is not provided.
- Adjust existing tests to pass the `childConfigDir` parameter where necessary.
- Add `afterEach` to restore all mocks after each test.

kapral18 committed:
chore: simplify ESLint no-restricted-imports configuration

- Refactor ESLint configurations to use `createNoRestrictedImportsOverride` for consistent restricted imports handling.
- Remove redundant type definitions and dynamic configuration merging.
- Streamline the configuration to enhance readability and maintainability.

kapral18 committed:
fix: update README to clarify parsing and directory handling

- Parses `.eslintrc.js` instead of relying on git for root discovery
- Identifies existing `no-restricted-imports` rules in the root config overrides
- Adds additional restricted imports to the existing ones
- Returns only the overrides that apply to the current directory based on `childConfigDir`
- Maintains all other ESLint settings from the root config
- Updated API documentation to include `childConfigDir` parameter
- Removed outdated notes about git repository requirement and root `.eslintrc.js` location

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/316045)
* Commit: e57591565c870c9e3081cca79f3db7646edaa15f

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/316045#0197e261-249c-46e5-bdb4-0ec9d4404bc0)

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/316045","number":316045,"commit":"e57591565c870c9e3081cca79f3db7646edaa15f"}],"number":316045}
buildkite-pr-comment-->

kapral18 committed:
fix: clone rootConfig to prevent mutation

- Clone `rootConfig` to `clonedRootConfig` to avoid mutating the original configuration.
- Update tests to verify that the original `rootConfig` is not mutated.
- Improve TODO comment in path mock for better clarity.

kapral18 commented on .github/CODEOWNERS:
```diff
@@ -61,6 +61,7 @@ packages/kbn-eslint-plugin-eui-a11y @elastic/obs-ux-infra_services-team
 packages/kbn-eslint-plugin-i18n @elastic/obs-knowledge-team @elastic/kibana-operations
 packages/kbn-eslint-plugin-imports @elastic/kibana-operations
 packages/kbn-eslint-plugin-telemetry @elastic/obs-knowledge-team
+packages/kbn-eslint-rule-overrides @elastic/kibana-operations
```
I originally thought to put threat-hunting as owners but I am transferring to platform team now and I feel like I can definitely use this in the new team as well, so instead kibana-operations looks like the team to own this.

But as I stated in the description I fully volunteer to help maintain this if you accept this into your files.

@elastic/kibana-operations

kapral18 committed:
chore: remove redundant comments

kapral18 commented on packages/kbn-eslint-rule-overrides/src/create_no_restricted_imports_override.js:
```diff
@@ -0,0 +1,227 @@
+const rootDir = path.resolve(__dirname, '..', '..', '..');
+
+// eslint-disable-next-line import/no-dynamic-require, @kbn/imports/no_boundary_crossing
+const rootConfig = require('../../../.eslintrc');
```
I understand we are crossing boundaries here by hardcoding the path to root eslint config file. But I'd argue we can't avoid it unless our root .eslintrc.js lives in its own package which would understandably break a lot of IDEs (unless symlinked). But either way it's not the case so for now this looks to me like a justified tradeoff. But I am open to suggestions.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/316508)
* Commit: 1ad1c398e9b7fafa760a1c49bf2c1a72faba0a8e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/945ef83e38843e6440fa7d05df9440e3304c1dc9) | [after](https://github.com/elastic/kibana/commit/1ad1c398e9b7fafa760a1c49bf2c1a72faba0a8e) | diff |
| --- | --- | --- | --- |
| `@kbn/cell-actions` | 3 | 2 | -1 |
| `@kbn/ecs-data-quality-dashboard` | 9 | 8 | -1 |
| `@kbn/eslint-rule-overrides` | - | 1 | +1 |
| `@kbn/security-solution-connectors` | 2 | 1 | -1 |
| `@kbn/security-solution-features` | 2 | 1 | -1 |
| `@kbn/security-solution-navigation` | 9 | 8 | -1 |
| `@kbn/security-solution-side-nav` | 4 | 3 | -1 |
| `@kbn/security-solution-upselling` | 3 | 2 | -1 |
| `@kbn/securitysolution-ecs` | 3 | 2 | -1 |
| `ecsDataQualityDashboard` | 7 | 6 | -1 |
| `securitySolution` | 629 | 624 | -5 |
| total |  |  | -13 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/945ef83e38843e6440fa7d05df9440e3304c1dc9) | [after](https://github.com/elastic/kibana/commit/1ad1c398e9b7fafa760a1c49bf2c1a72faba0a8e) | diff |
| --- | --- | --- | --- |
| `@kbn/cell-actions` | 3 | 2 | -1 |
| `@kbn/ecs-data-quality-dashboard` | 9 | 8 | -1 |
| `@kbn/eslint-rule-overrides` | - | 1 | +1 |
| `@kbn/security-solution-connectors` | 2 | 1 | -1 |
| `@kbn/security-solution-features` | 2 | 1 | -1 |
| `@kbn/security-solution-navigation` | 9 | 8 | -1 |
| `@kbn/security-solution-side-nav` | 4 | 3 | -1 |
| `@kbn/security-solution-upselling` | 3 | 2 | -1 |
| `@kbn/securitysolution-ecs` | 3 | 2 | -1 |
| `ecsDataQualityDashboard` | 7 | 6 | -1 |
| `securitySolution` | 727 | 722 | -5 |
| total |  |  | -13 |


</details>


### History
* :broken_heart: [Build #316225](https://buildkite.com/elastic/kibana-pull-request/builds/316225) failed b3094f8494357f36c6375a9191f131e3953b6753
* :broken_heart: [Build #316045](https://buildkite.com/elastic/kibana-pull-request/builds/316045) failed e57591565c870c9e3081cca79f3db7646edaa15f

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/316508","number":316508,"commit":"1ad1c398e9b7fafa760a1c49bf2c1a72faba0a8e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/316225","number":316225,"commit":"b3094f8494357f36c6375a9191f131e3953b6753"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/316045","number":316045,"commit":"e57591565c870c9e3081cca79f3db7646edaa15f"}],"number":316508}
buildkite-pr-comment-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

kapral18 committed:
chore(deps): update minimatch to 8, add excludedFiles support & more

- Update minimatch to version ^8.0.4 in `package.json` to use hasMagic,
  braceExpand and other functionality that's available in 7+
- Add support for overrides[].excludedFiles
- Add support for complex partial and negated glob patterns
- Refactor `create_no_restricted_imports_override.js` to `index.js` with improved pattern transformation.
- Add `index.test.js` for testing the refactored override creation.
- Introduce utility functions `literal_path_transformer.js` and `pattern_transformer.js` with respective tests.
- Introduce `segment_matcher.js` with tests to handle pattern matching and segment consumption.
- Adjust `yarn.lock` to reflect the new version of minimatch and updated dependencies.

kapral18 committed:
fix: update `transformFilePatterns` to handle negated file patterns correctly and add type definitions

- Updated `transformFilePatterns` to correctly handle negated patterns, especially when they exclude the target directory.
- Added type definitions for `RestrictedImportString`, `RestrictedImportPath`, and `CreateOverrideOptions` to improve code clarity and maintainability.
- Enhanced the `transformFilePatterns` function to differentiate between `files` and `excludedFiles` patterns properly.
- Modified the `index.test.js` file to include tests for negated patterns, including both simple and complex brace patterns.
- Added error handling and validation for invalid `restrictedImports` values.
- Ensured that malformed patterns and empty file patterns are gracefully handled, returning valid results.

kapral18 committed:
chore: add minimatch upgrade related changes

kapral18 committed:
fix: correct type annotation for RESTRICTED_IMPORTS_PATHS

- Updated the type annotation from `Array.<RestrictedImportPath>` to `RestrictedImportPath[]` in multiple `.eslintrc.js` files.
- Corrected variable name from `RESTRICTED_IMPORTS` to `RESTRICTED_IMPORTS_PATHS` in `x-pack/solutions/security/plugins/security_solution/public/onboarding/.eslintrc.js`.

kapral18 committed:
refactor: modularize no-restricted-imports override creation

- **Refactor** the `createNoRestrictedImportsOverride` function to modularize the process of merging restricted imports and transforming file patterns.
- **Remove** the `pattern_transformer.js` and its tests as their functionality was split into `transform_patterns.js`.
- **Create** `merge_restricted_imports.js` to handle merging restricted imports into existing ESLint rules.
- **Create** `transform_overrides_for_nested_context.js` to handle transforming ESLint overrides for nested contexts.
- **Create** `transform_patterns.js` to handle the transformation of file and excluded files patterns.
- **Update** tests to reflect changes in the refactored code, ensuring comprehensive coverage of new utility functions and the main function.
- **Ensure** that the original root config is not mutated by working with a deep clone.
- **Add** detailed comments and type definitions for better readability and maintainability.

kapral18 committed:
feat: enhance README with detailed documentation for @kbn/eslint-rule-overrides

- Update purpose section to focus on `no-restricted-imports` rules
- Add "How It Works" section detailing key transformations
- Include pattern transformation examples and behavior
- Revise installation instructions for clarity
- Expand usage examples with basic and advanced scenarios
- Add API reference for `createNoRestrictedImportsOverride`
- Include parameters, return values, and error handling
- Detail pattern transformation behavior for files and excludedFiles
- Explain deduplication behavior for conflicting restrictions
- Provide real-world example for better understanding
- Add troubleshooting section with common issues and debug tips
- Include contributing guidelines and license information

kapral18 committed:
improvement: unify createRuleOverrides interface

- Extend README to cover new features and strategies
- Refactor `createNoRestrictedImportsOverride` to `createRuleOverrides`
- Implement new strategies: merge, replace, remove, append, prepend
- Create rule handlers for customizable rule processing
- Add tests for new functionality and strategies
- Update documentation and examples accordingly
- Move `mergeRestrictedImports` to `no-restricted-imports` handler
- Enhance error handling and validation for rule configurations

kapral18 committed:
fix: Restructure README and enhance createRuleOverrides functionality

- Restructured README to improve clarity and organization
- Added detailed sections for usage, strategies, and contributing handlers
- Enhanced `createRuleOverrides` to support severity-only changes and improved error handling
- Updated `normalizeRuleConfig` to handle severity-only configurations
- Modified `applyDefaultHandler` to support severity-only changes and enforce custom handlers for merge, append, prepend strategies
- Added comprehensive tests for new features and improvements

kapral18 committed:
fix: update .eslintrc.js files to use createRuleOverrides

- Updated the import from `createNoRestrictedImportsOverride` to `createRuleOverrides` in multiple `.eslintrc.js` files.
- Modified the configuration to use the new `rules` structure with `strategy` set to `append` for `no-restricted-imports`.

kapral18 committed:
test: bdd-fy tests

kapral18 committed:
improve: simplify strategies, remove append, prepend

- Removed `append` and `prepend` strategies from the README and codebase.
- Updated `create_rule_overrides.js` to remove validation and handling for `append` and `prepend`.
- Updated `create_rule_overrides.test.js` to remove tests related to `append` and `prepend`.
- Updated `no-restricted-imports/index.js` to remove `append` and `prepend` logic.
- Updated `no-restricted-imports/index.test.js` to remove tests related to `append` and `prepend`.

kapral18 committed:
chore: intermediate ideas
